### PR TITLE
855 grouping legal graph

### PR DESCRIPTION
**Pull request by ShiriMoran**

### CODE REVIEW

**Review by zivnevo — CHANGES_REQUESTED** [review]
Make grouping an enum - NoGrouping, Grouping, GroupingWithConsistencyEdges

File: pkg/vpcmodel/groupingGraphical.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+import (
+	"github.com/np-guard/models/pkg/ipblock"
+)
+
+// for the graphical (html, drawio, svg) representation. In the graph presentation, each node must have all relevant edges.
+// this is not the case in the textual presentation. E.g., a textual presentation may look like:
+// 142.0.64.0/17 ->vsi2
+// 142.0.0.0/16 -> vsi1
+// 0.0.0.0/0 -> vsi3
+// 142.0.64.0/17 should also be connected to vsi2 and vsi3
```
Comment (zivnevo) [inline pkg/vpcmodel/groupingGraphical.go:18]:
```suggestion
// 142.0.64.0/17 should also be connected to vsi1 and vsi3
```

**Comment by ShiriMoran** [comment]
> GroupingWithConsistencyEdges

And NoGroupingWithConsistencyEdges

File: pkg/vpcmodel/groupingGraphical.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+
+func groupedExternalToIPBlock(ee EndpointElem) *ipblock.IPBlock {
+	// EndpointElem must be of type groupedExternalNodes
+	elements := []*ExternalNetwork(*ee.(*groupedExternalNodes))
+	var res = ipblock.New()
+	for _, e := range elements {
+		res = res.Union(e.ipblock)
+	}
+	return res
+}
```
Comment (zivnevo) [inline pkg/vpcmodel/groupingGraphical.go:74]:
```suggestion
func groupedExternalToIPBlock(ee *groupedExternalNodes) *ipblock.IPBlock {
	var res = ipblock.New()
	for _, e := range *ee {
		res = res.Union(e.ipblock)
	}
	return res
}
```
Better enforcement of parameter type

File: pkg/vpcmodel/groupingGraphical.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+// In order to add missing edges, we go over all the endpoints in grouping that present external nodes, and check for containment
+// if external endpoint e1 is contained in external end point e2 then all the "edges" of e2 are added to e1
+func (g *GroupConnLines) consistencyEdgesExternal() {
+	// 1. Gets a map from external endpoints name to their IPs
+	eeNameToIPBlock := getMapToIps(g.GroupedLines)
+	// 2. Gets a map from external endpoints name to their endpoint
+	eeNameToEE := getMapToEPEs(g.GroupedLines)
+	// 3. Gets a map from external endpoint name to all the endpoint it contains
+	containedMap := getContainedEndpointMap(eeNameToIPBlock, eeNameToEE)
+	// 3. Add edges, based on the above map (3)
```
Comment (zivnevo) [inline pkg/vpcmodel/groupingGraphical.go:28]:
```suggestion
	// 4. Add edges, based on the above map (3)
```

File: pkg/vpcmodel/groupingGraphical.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+	g.addEdgesOfContainingEPs(containedMap)
+}
+
+// gets []*groupedConnLine and returns a map from the string presentation of each endpoint to its ipBlock
+func getMapToIps(grouped []*groupedConnLine) (eeToIPBlock map[string]*ipblock.IPBlock) {
```
Comment (zivnevo) [inline pkg/vpcmodel/groupingGraphical.go:33]:
```suggestion
func getMapToIps(grouped []*groupedConnLine) (map[string]*ipblock.IPBlock) {
```
Naming the returned value doesn't help much when its type is a map (also in other functions in this file)

Comment (ShiriMoran) [reply pkg/vpcmodel/groupingGraphical.go:33]: why?

File: pkg/vpcmodel/groupingGraphical.go (hunk 2 above)
Comment (ShiriMoran) [reply pkg/vpcmodel/groupingGraphical.go]: done (required change a bit broader than the suggested)

**Review by zivnevo — APPROVED** [review]

**Review by haim-kermany — APPROVED** [review]

**Pull request merged**